### PR TITLE
Fix some polling/StrictMode issues, along with some general refactorings

### PR DESCRIPTION
**Pull request by brainkim**
### Checklist:

- [x] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### Notes
Continuation of https://github.com/apollographql/apollo-client/pull/8346

Spent some time deleting Reobserver, as well as fixing some polling misbehavior.

Most of the test changes are related to fixing StrictMode stuff. The timings of moving ObservableQuery logic to useEffect made it so that we need little bits of delay to get the tests working. The behavior doesn’t seem to have changed in any meaningful way, we just skip some renders when we refetch or rerender too soon. There was one test which just didn’t make sense, was never failable in any meaningful sense, and used timer mocks in a way which messed with other tests so I deleted it. Forgive me.

This should probably make Apollo Client “strict mode” safe, though it is roughly impossible to tell for certain.

### Fixes
- https://github.com/apollographql/apollo-client/issues/6391
- https://github.com/apollographql/apollo-client/issues/7221
- https://github.com/apollographql/apollo-client/issues/6391

### TODO: Investigate if this fixes
- ~~https://github.com/apollographql/apollo-client/issues/6858~~
- https://github.com/apollographql/apollo-client/issues/6661
- https://github.com/apollographql/apollo-client/issues/5870
- https://github.com/apollographql/apollo-client/issues/5140

### CODE REVIEW

File: src/react/data/QueryData.ts [hunk 1]
```diff
@@ -151,9 +156,7 @@ export class QueryData<TData, TVariables> extends OperationData<
   };
 
   private getExecuteResult(): QueryResult<TData, TVariables> {
-    const result = this.getQueryResult();
-    this.startQuerySubscription();
-    return result;
+    return this.getQueryResult();
```
Comment (benjamn) [inline src/react/data/QueryData.ts:159]: Maybe get rid of `getExecuteResult` and just call `this.getQueryResult()` instead?

File: src/react/components/__tests__/client/Query.test.tsx [hunk 2]
```diff
@@ -1122,98 +1118,6 @@ describe('Query component', () => {
       return wait(() => expect(count).toBe(2)).then(resolve, reject);
     });
 
-    itAsync('use client from props or context', (resolve, reject) => {
-      jest.useFakeTimers();
```
Comment (benjamn) [inline src/react/components/__tests__/client/Query.test.tsx:1126]: I support the solution of simply removing this test.

File: src/core/ObservableQuery.ts [hunk 3]
```diff
@@ -537,19 +532,13 @@ once, rather than every time you call fetchMore.`);
...
 
   public stopPolling() {
-    this.updateOptions({ pollInterval: 0 });
-  }
-
-  public updateOptions(
-    newOptions: Partial<WatchQueryOptions<TVariables, TData>>,
-  ) {
-    Object.assign(this.options, compact(newOptions));
+    this.options.pollInterval = 0;
```
Comment (benjamn) [inline src/core/ObservableQuery.ts]: This resolves my one concern with f0639578101b430c0141dae3c90a077d48cc2121, which was whether `updateOptions` (and `stop` below) really needed to be `public`.

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 4]
```diff
@@ -720,6 +720,69 @@ describe('useQuery Hook', () => {
...
+
+          switch (++renderCount) {
+            case 1:
+            case 2:
+              expect(loading).toBeTruthy();
+              expect(data).toBeUndefined();
+              break;
+            case 3:
+            case 4:
+              expect(loading).toBeFalsy();
```
Comment (benjamn) [inline src/react/hooks/__tests__/useQuery.test.tsx:761]: I can confirm this test fails as expected (on this line) when cherry-picked onto `release-3.4`, without including the rest of this PR.

File: src/core/ObservableQuery.ts [hunk 5]
```diff
@@ -91,12 +131,14 @@ export class ObservableQuery<
...
     this.queryManager = queryManager;
-
     this.queryInfo = queryInfo;
   }
 
   public result(): Promise<ApolloQueryResult<TData>> {
     return new Promise((resolve, reject) => {
+      // TODO: this code doesn’t actually make sense insofar as the observer
+      // will never exist in this.observers due how zen-observable wraps observables.
+      // https://github.com/zenparsing/zen-observable/blob/master/src/Observable.js#L169
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:141]: How should we address this TODO?

File: src/react/hoc/__tests__/mutations/queries.test.tsx [hunk 6]
```diff
@@ -281,13 +281,15 @@ describe('graphql(mutation) query integration', () => {
...
-              .then(result => {
-                expect(stripSymbols(result && result.data)).toEqual(
-                  mutationData
-                );
-              })
-              .catch(reject);
+            setTimeout(() => {
+              this.props.mutate!()
+                .then(result => {
+                  expect(stripSymbols(result && result.data)).toEqual(
```
Comment (benjamn) [inline src/react/hoc/__tests__/mutations/queries.test.tsx:287]: Just a note (not relevant to this commit) that I'm fairly certain we no longer need to use `stripSymbols` (at least in future tests, but probably not in legacy tests either).

File: src/core/ObservableQuery.ts [hunk 7]
```diff
@@ -98,13 +98,7 @@ export class ObservableQuery<
 
     // related classes
     this.queryManager = queryManager;
-
     this.queryInfo = queryInfo;
-
-    // Avoid polling during SSR and when the query is already in flight.
-    this.shouldFetch =
-      !this.queryManager.ssrMode &&
-      (() => !isNetworkRequestInFlight(this.queryInfo.networkStatus));
```
Comment (benjamn) [inline src/core/ObservableQuery.ts]: Nice to see this simplified away!

**Review by benjamn — APPROVED** [review]
Thanks to the granularity of your commits, @brainkim, these changes were actually pretty easy to review! I left a few comments in-line, but nothing merge-blocking.

Speaking of merging, feel free to rebase and squash commits together before merging if you like, but I think I'd prefer to keep these separate commits (with a merge commit), instead of squashing the whole PR.

File: src/core/ObservableQuery.ts (hunk 5 above)
Comment (brainkim) [reply src/core/ObservableQuery.ts:141]: Not sure yet. I have to understand how the `result()` method is used in the wild, and the context of the comments, but we can climb this hill later I think.

**Review by hwillson — APPROVED** [review]
This looks great @brainkim - thanks very much!

(just a gentle reminder to update the `CHANGELOG` 👍)

**Comment by benjamn** [comment]
@brainkim I just released this in `@apollo/client@3.4.0-rc.13`, fyi.

**Pull request merged**